### DISCUSSION TITLE
Improve LOD to account for camera parameters

### DISCUSSION THREAD

**Discussion by KaruroChori** [godotengine/godot-proposals#4659] [3D]
As far as I know the current implementation for LODs in godot only uses distance from the active camera to determine the degree of simplification in the mesh. However this does not account for focal length, which has a major impact as well.
I would suggest to integrate this metric as well in the evaluation.

**Comment by KaruroChori** [comment]
My assumptions were wrong.
`CameraMatrix::get_lod_multiplier()` is used to generate a factor whic is later used in `surf->sort.lod_index = mesh_storage->mesh_surface_get_lod(surf->surface, inst->lod_model_scale * inst->lod_bias, distance * p_render_data->lod_distance_multiplier, p_render_data->screen_mesh_lod_threshold, &indices);` as lod_distance_multiplier.

**Comment by Calinou** [comment]
Automatic mesh LOD already takes the camera FOV into account, simply because it uses a screen density metric instead of distance between the object and the camera. This also means it adapts to viewport resolution changes.

On the other hand, visibility ranges (manual HLOD) do not take the camera FOV or viewport resolution into account. The distance thresholds set by the user remain fixed no matter what. I think adapting to FOV makes sense, but this will require defining a "LOD base camera FOV" in the project settings in case someone is using a camera FOV other than 75 as a default value. Otherwise, distance thresholds in the editor won't match up with in-game distance thresholds. Adapting to screen resolution is also possible, but it can make adjustments difficult to understand, especially if this is also done with LOD FOV adjustment at the same time. So I'd prefer we stick to LOD FOV adjustments for now.

For reference, this was also discussed in https://github.com/godotengine/godot-proposals/discussions/4204.

**Reply by KaruroChori** [reply]
Thank you for the reply; this explains the inconsistent behaviour I was observing.
